8314684: Add overview docs to loaderConstraints.cpp

iklam opened the pull request:
I added some documentation to make it easier to understand loaderConstraints.cpp. Please review and check if my description is actually correct. Thanks!

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314684](https://bugs.openjdk.org/browse/JDK-8314684): Add overview docs to loaderConstraints.cpp (**Enhancement** - P4)


### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15372/head:pull/15372` \
`$ git checkout pull/15372`

Update a local copy of the PR: \
`$ git checkout pull/15372` \
`$ git pull https://git.openjdk.org/jdk.git pull/15372/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15372`

View PR using the GUI difftool: \
`$ git pr show -t 15372`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15372.diff">https://git.openjdk.org/jdk/pull/15372.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15372#issuecomment-1687061373)

calvinccheung commented on src/hotspot/share/classfile/loaderConstraints.cpp:
```diff
@@ -38,7 +38,47 @@
+//
+// Management of LoaderConstraints
+//
+// When the SyswtemDictionary decides that loader_x and loader_y must resolve the name N to the same class:
```
Typo: SyswtemDictionary should be SystemDictionary

calvinccheung commented on src/hotspot/share/classfile/loaderConstraints.cpp:
```diff
@@ -38,7 +38,47 @@
+// we assign _klass_<m> = K;
+//
+// if _klass_<m> is non-null, when a loader loader_<m><n> tries to resolve the name N to a class K,
+// where _klass_<m> != K, a LinkageError is thrown, and the resolution fails;
```
Suggestion: '.' instead of ';'

calvinccheung commented on src/hotspot/share/classfile/loaderConstraints.cpp:
```diff
@@ -38,7 +38,47 @@
+// both loader_x and loader_y;
+//
+// Otherwise if LoaderConstraint_x exists but LoaderConstraint_y doesn't exist, add loader_y to LoaderConstraint_x;
+// (and vice versa);
```
Suggestion: s/and/or
Also, I think the parentheses could be removed.

calvinccheung reviewed (approved):
Looks good. Just noticed some typos.

dholmes-ora reviewed (commented):
Well it reads really well and is informative, but is it correct? Hard to evaluate. It would be clearer if the commentary tied into the code more so that I knew where to see each of the things mentioned, and also understand how the different types (LoaderConstraint, ConstraintSet) fit in that description.

Thanks.

iklam commented:
> Well it reads really well and is informative, but is it correct? Hard to evaluate. It would be clearer if the commentary tied into the code more so that I knew where to see each of the things mentioned, and also understand how the different types (LoaderConstraint, ConstraintSet) fit in that description.
> 
> Thanks.

I added ConstraintSet to the comments. I also added references to the `add_entry` and `check_or_update` functions which are the main entry points.

dholmes-ora commented on src/hotspot/share/classfile/loaderConstraints.cpp:
```diff
@@ -38,7 +38,50 @@
+//
+// Otherwise if both LoaderConstraints have different values for _klass, a LinkageError is thrown.
+//
+// Otherwise the two LoaderConstraints are merged into one.
```
So I looked in `LoaderConstraintTable::add_entry` to see how this algorithm applied and I could not figure out what condition is actually left (regarding `pp1` and `pp2`) when we get to the merge case at L396 ???

dholmes-ora replied:
Never mind I see it now: `pp1` != `pp2` and neither is null.

dholmes-ora reviewed (approved):
Thanks for the update.

Looks good - very useful.

iklam commented:
Thanks for the review @calvinccheung and @dholmes-ora 
/integrate
